### PR TITLE
Fix corner case with dead server coming back alive

### PR DESCRIPTION
**Pull request by scoiatael**
Hey, 

I haven't opened an issue for that as I believe it's easier to show failing tests.

Looks like `clients/hash` behaves incorrectly when dead server comes back alive - it's never removed from `dead_clients` and thus retried indefinitely. 

Failing test and fix are visible in 35ee9de, rest of PR simply adds some test coverage and refactors existing code a little.

### CODE REVIEW

File: pymemcache/client/hash.py [hunk 1]
```diff
@@ -121,22 +121,29 @@ def remove_server(self, server, port):
...
+                    'bringing server back into rotation %s',
+                    server
+                )
+                self.add_server(*server)
+                del self._dead_clients[server]
+            self._last_dead_check_time = current_time
+
     def _get_client(self, key):
         check_key_helper(key, self.allow_unicode_keys, self.key_prefix)
         if len(self._dead_clients) > 0:
```
Comment (jparise) [inline pymemcache/client/hash.py]: While we're here, this line could be simplified:

```suggestion
        if self._dead_clients):
```

File: pymemcache/client/hash.py [hunk 2]
```diff
@@ -121,22 +121,29 @@ def remove_server(self, server, port):
         key = '%s:%s' % (server, port)
         self.hasher.remove_node(key)
 
+    def _retry_dead(self):
+        current_time = time.time()
+        ldc = self._last_dead_check_time
+        # we have dead clients and we have reached the
+        # timeout retry
```
Comment (jparise) [inline pymemcache/client/hash.py]: Note that the guarantee that we have dead clients is now external to this function.

**Comment by jparise** [comment]
@scoiatael, just checking in on this. Did you comments make sense?

**Comment by scoiatael** [comment]
Yeah, thanks for them. Simply didn't have time recently to apply them.

Best regards,

Łukasz Czaplinski

File: pymemcache/client/hash.py (hunk 2 above)
Comment (scoiatael) [reply pymemcache/client/hash.py:128]: That's one way of viewing it. The other is that this function works regardless of whether we have dead clients or not - meaning the check for empty clients is simply an optimisation.

I'll update the comment accordingly.

**Comment by jparise** [comment]
Thank you, @scoiatae!
